From a1dd2584d3fa86c0943d9fb02191d746d15515c3 Mon Sep 17 00:00:00 2001
From: zhangmeng <zhangmeng.kevin@linux.spacemit.com>
Date: Mon, 25 Aug 2025 19:19:57 +0800
Subject: [PATCH 39/43] riscv: k1: dts: enable sdhci-0/1/2 for bananapi f3
 board

community inclusion
category: feature
bugzilla: https://github.com/RVCK-Project/rvck-olk/issues/114

--------------------------------

Three sdhci controllers are enabled on bananapi-f3 board and used as:
- sdhci-0: enabled as a tf-card slot
- sdhci-1: enabled by sdio-wifi (rtl8852bs module)
- sdhci-2: enabled by emmc

Signed-off-by: zhangmeng <zhangmeng.kevin@linux.spacemit.com>
---
 .../boot/dts/spacemit/k1-bananapi-f3.dts      | 66 +++++++++++++++++++
 1 file changed, 66 insertions(+)

diff --git a/arch/riscv/boot/dts/spacemit/k1-bananapi-f3.dts b/arch/riscv/boot/dts/spacemit/k1-bananapi-f3.dts
index 74adf460fccc3..2fc4e4cf52dab 100644
--- a/arch/riscv/boot/dts/spacemit/k1-bananapi-f3.dts
+++ b/arch/riscv/boot/dts/spacemit/k1-bananapi-f3.dts
@@ -310,3 +310,69 @@ &pwm7 {
 	pinctrl-0 = <&pinctrl_pwm7_0>;
 	status = "okay";
 };
+
+/* SDCard */
+&sdhci0 {
+	pinctrl-names = "default","fast";
+	pinctrl-0 = <&pinctrl_mmc1>;
+	pinctrl-1 = <&pinctrl_mmc1_fast>;
+	bus-width = <4>;
+	cd-gpios = <&gpio 80 0>;
+	cd-inverted;
+	vmmc-supply = <&dcdc_4>;
+	vqmmc-supply = <&ldo_1>;
+	no-mmc;
+	no-sdio;
+	spacemit,sdh-host-caps-disable = <(MMC_CAP_UHS_SDR12 | MMC_CAP_UHS_SDR25)>;
+	spacemit,sdh-quirks = <(SDHCI_QUIRK_BROKEN_CARD_DETECTION |
+				SDHCI_QUIRK_INVERTED_WRITE_PROTECT |
+				SDHCI_QUIRK_BROKEN_TIMEOUT_VAL)>;
+	spacemit,sdh-quirks2 = <(SDHCI_QUIRK2_PRESET_VALUE_BROKEN |
+				 SDHCI_QUIRK2_BROKEN_PHY_MODULE |
+				 SDHCI_QUIRK2_SET_AIB_MMC)>;
+	spacemit,aib_mmc1_io_reg = <0xD401E81C>;
+	spacemit,apbc_asfar_reg = <0xD4015050>;
+	spacemit,apbc_assar_reg = <0xD4015054>;
+	spacemit,rx_dline_reg = <0x0>;
+	spacemit,tx_dline_reg = <0x0>;
+	spacemit,tx_delaycode = <0x5f>;
+	spacemit,rx_tuning_limit = <50>;
+	spacemit,sdh-freq = <204800000>;
+	status = "okay";
+};
+
+/* SDIO */
+&sdhci1 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_mmc2>;
+	bus-width = <4>;
+	non-removable;
+	vqmmc-supply = <&dcdc_3>;
+	no-mmc;
+	no-sd;
+	spacemit,sdh-host-caps-disable = <(MMC_CAP_UHS_DDR50 | MMC_CAP_NEEDS_POLL)>;
+	spacemit,sdh-quirks = <(SDHCI_QUIRK_BROKEN_CARD_DETECTION |
+				SDHCI_QUIRK_BROKEN_TIMEOUT_VAL)>;
+	spacemit,sdh-quirks2 = <(SDHCI_QUIRK2_PRESET_VALUE_BROKEN |
+				 SDHCI_QUIRK2_BROKEN_PHY_MODULE)>;
+	spacemit,rx_dline_reg = <0x0>;
+	spacemit,tx_delaycode = <0x8f 0x5f>;
+	spacemit,rx_tuning_limit = <50>;
+	spacemit,sdh-freq = <375000000>;
+	status = "okay";
+};
+
+/* eMMC */
+&sdhci2 {
+	bus-width = <8>;
+	non-removable;
+	mmc-hs400-1_8v;
+	mmc-hs400-enhanced-strobe;
+	no-sd;
+	no-sdio;
+	spacemit,sdh-quirks = <(SDHCI_QUIRK_BROKEN_CARD_DETECTION |
+				SDHCI_QUIRK_BROKEN_TIMEOUT_VAL)>;
+	spacemit,sdh-quirks2 = <SDHCI_QUIRK2_PRESET_VALUE_BROKEN>;
+	spacemit,sdh-freq = <375000000>;
+	status = "okay";
+};
-- 
2.51.0

